### ISSUE TITLE
diffChangeLog constantly creating and dropping indicies/primary keys since Spring Boot 3/Liquibase 4.19

### ISSUE THREAD

**Issue by Lukas0610** [liquibase/liquibase-hibernate#456]
Since upgrading to Spring Boot 3 (and thus to Liquibase 4.19), running diffChangeLog using `hibernate:spring:*` as a reference, Liquibase generates changelogs first dropping a constraint and then re-creates it with the same columns right after the drop-statement:

``` xml
<changeSet author="root (generated)" id="1674298858677-83">
    <dropUniqueConstraint constraintName="UC_ACCOUNTSTRIPE_CUSTOMER_ID_COL" tableName="account"/>
</changeSet>
<changeSet author="root (generated)" id="1674298858677-84">
    <addUniqueConstraint columnNames="stripe_customer_id" constraintName="UC_ACCOUNTSTRIPE_CUSTOMER_ID_COL" tableName="account"/>
</changeSet>
```

A similar thing happens to Envers audit-tables, just with the primary key in that case:

``` xml
<changeSet author="root (generated)" id="1674298858677-81">
    <dropPrimaryKey tableName="$audit_stop_groups"/>
</changeSet>
<changeSet author="root (generated)" id="1674298858677-82">
    <addPrimaryKey columnNames="stop_id, groups_id, revision" constraintName="$audit_stop_groupsPK" tableName="$audit_stop_groups"/>
</changeSet>
```

When executing either of these combined change-set, on the next run of diffChangeLog, the same changesets are generated again.

Another thing is that for the same Envers audit-tables, first an `createIndex`-changeset is being created. After executing these and running diffChangeLog again, a `dropIndex`-changesets for the just created indices are being created.
``` xml
<changeSet author="root (generated)" id="1674298858677-183">
    <createIndex indexName="IX_$audit_stop_groupsPK" tableName="$audit_stop_groups" unique="true">
        <column name="stop_id"/>
        <column name="groups_id"/>
        <column name="revision"/>
    </createIndex>
</changeSet>
```
From what I can see, these changesets are only created for ManyToMany-relations and ElementCollection-annotated lists.

**Comment by fleboulch** [comment]
Related to #436

**Comment by amazing4u** [comment]
This problem is fixed with version 4.21.0

**Comment by fleboulch** [comment]
Thanks a lot !

**Comment by grigala** [comment]
> This problem is fixed with version 4.21.0

It wasn't in my case 🤷‍♂️  I [still ](https://github.com/liquibase/liquibase-hibernate/issues/456#issuecomment-1400803652)get `dropUniqueConstraint`, `addUniqueConstraint` list of changes.
- `org.liquibase:liquibase-core:4.21.1`
- `org.liquibase.ext:liquibase-hibernate5:4.21.1`

```xml
 <changeSet author="george (generated)" id="1674498709369-1">
        <dropUniqueConstraint constraintName="UC_APPLICATION_PRINCIPALID_COL" tableName="...">
    </changeSet>
    <changeSet author="george (generated)" id="1674498709369-2">
        <addUniqueConstraint columnNames="id" constraintName="UC_APPLICATION_PRINCIPALID_COL" tableName="..."/>
    </changeSet>
...
```

**Comment by amazing4u** [comment]
> > This problem is fixed with version 4.21.0
> 
> It wasn't in my case 🤷‍♂️ I [still ](https://github.com/liquibase/liquibase-hibernate/issues/456#issuecomment-1400803652)get `dropUniqueConstraint`, `addUniqueConstraint` list of changes.
> 
> * `org.liquibase:liquibase-core:4.21.1`
> * `org.liquibase.ext:liquibase-hibernate5:4.21.1`
> 
> ```
>  <changeSet author="george (generated)" id="1674498709369-1">
>         <dropUniqueConstraint constraintName="UC_APPLICATION_PRINCIPALID_COL" tableName="...">
>     </changeSet>
>     <changeSet author="george (generated)" id="1674498709369-2">
>         <addUniqueConstraint columnNames="id" constraintName="UC_APPLICATION_PRINCIPALID_COL" tableName="..."/>
>     </changeSet>
> ...
> ```

You are right. Sorry. I mixed it up with another bug. The problem you described was fixed today: 
https://github.com/liquibase/liquibase-hibernate/issues/479
https://github.com/liquibase/liquibase-hibernate/pull/480

Hopefully it will be released soon

**Comment by grigala** [comment]
> 
> You are right. Sorry. I mixed it up with another bug. The problem you described was fixed today: #479 #480
> 
> Hopefully it will be released soon

Thanks for the hints. Yeah, I certainly hope so. 

They have also introduced a breaking change somewhere along the line with `CommandScope` and `DiffChangelogCommandStep`. Namely, executing `diffChangelog` command scope and watching for results isn't working the same way it used to.

```java
CommandResults results = new CommandScope("diffChangelog").addArgumentValue(DiffChangelogCommandStep.CHANGELOG_FILE_ARG, outputFile)
                                                          .addArgumentValue(DiffChangelogCommandStep.URL_ARG, URL)
                                                          .addArgumentValue(DiffChangelogCommandStep.USERNAME_ARG, USERNAME)
                                                          .addArgumentValue(DiffChangelogCommandStep.PASSWORD_ARG, PASSWORD)
                                                          .addArgumentValue(DiffChangelogCommandStep.REFERENCE_URL_ARG, REFERENCE_URL)
                                                          .addArgumentValue(DiffChangelogCommandStep.EXCLUDE_OBJECTS_ARG, EXCLUDED_OBJECTS)
                                                          .execute();
```

This will not work any more as they've [removed a lot of static variables](https://github.com/liquibase/liquibase/commit/6be41a80cee2ea3fddc5fe8b10d08552cf8e4538#diff-06e1b1f8b18c1679dbf1305e3bd6b67c5e30f73ff9a89bf961a179387e86e285) from the `DiffChangelogCommandStep` class, so now I'm guessing you have to do this to achieve the same result?
```java
CommandResults results = new CommandScope(DiffChangelogCommandStep.COMMAND_NAME)
                .addArgumentValue(DiffChangelogCommandStep.CHANGELOG_FILE_ARG, outputFile)
                .addArgumentValue(PreCompareCommandStep.EXCLUDE_OBJECTS_ARG, EXCLUDED_OBJECTS)
                .addArgumentValue(GenerateChangelogCommandStep.REFERENCE_URL_ARG, REFERENCE_URL)
                .addArgumentValue(DbUrlConnectionCommandStep.URL_ARG, URL)
                .addArgumentValue(DbUrlConnectionCommandStep.USERNAME_ARG, USERNAME)
                .addArgumentValue(DbUrlConnectionCommandStep.PASSWORD_ARG,  PASSWORD)
                .execute();
```

Notice usages of `DbUrlConnectionCommandStep`, `DbUrlConnectionCommandStep` and `PreCompareCommandStep`.

The changes have been made [here(6be41a8)](https://github.com/liquibase/liquibase/commit/6be41a80cee2ea3fddc5fe8b10d08552cf8e4538) but I have no idea what was the reason to refactor it or what's the new way of doing the same as I couldn't find any documentation related to it.
